File: code/library/src/main/java/org/fbme/lib/iec61499/stringify/SubappNetworkPrinter.java
package org.fbme.lib.iec61499.stringify;


import org.fbme.lib.iec61499.fbnetwork.subapp.SubappNetwork;
import org.fbme.lib.iec61499.fbnetwork.subapp.SubapplicationDeclaration;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;

public class SubappNetworkPrinter extends FBNetworkPrinter<SubappNetwork> {

    public SubappNetworkPrinter(SubappNetwork network) {
        super(network, "SubAppNetwork");
    }

    @Override
    protected void printElementBody(Element element) {
        for (SubapplicationDeclaration declaration : myElement.getSubapplications()) {
            element.addContent(new SubapplicationPrinter(declaration).print());
        }
        super.printElementBody(element);
    }

    private static class SubapplicationPrinter extends DeclarationPrinterBase<SubapplicationDeclaration> {

        public SubapplicationPrinter(SubapplicationDeclaration declaration) {
            super(declaration, "SubApp");
        }

        @Override
        protected void printDeclarationBody(@NotNull Element element) {
            element.setAttribute("Type", myElement.getTypeReference().getPresentation());
            element.setAttribute("x", "" + myElement.getX());
            element.setAttribute("y", "" + myElement.getY());
        }
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/stringify/SubappNetworkPrinter.kt
package org.fbme.lib.iec61499.stringify

import org.fbme.lib.iec61499.fbnetwork.subapp.SubappNetwork
import org.fbme.lib.iec61499.fbnetwork.subapp.SubapplicationDeclaration
import org.jdom.Element

class SubappNetworkPrinter(network: SubappNetwork) : FBNetworkPrinter<SubappNetwork>(network, "SubAppNetwork") {
    override fun printElementBody(element: Element) {
        for (declaration in this.element.subapplications) {
            element.addContent(SubapplicationPrinter(declaration).print())
        }
        super.printElementBody(element)
    }

    private class SubapplicationPrinter(declaration: SubapplicationDeclaration) :
        DeclarationPrinterBase<SubapplicationDeclaration>(declaration, "SubApp") {
        override fun printDeclarationBody(element: Element) {
            element.setAttribute("Type", this.element.typeReference.presentation)
            element.setAttribute("x", "" + this.element.x)
            element.setAttribute("y", "" + this.element.y)
        }
    }
}
